File: 560-subarray-sum-equals-k/subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
    sum := 0
    all_sum := map[int]int{
        0 : 1,
    }
    subarray := 0
    for i := range nums {
        sum += nums[i]
        complement := sum - k

        if val, ok := all_sum[complement]; ok {
            subarray += val
        }

        all_sum[sum] += 1
    }

    return subarray
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> prefix;
        prefix[0] = 1;
        int sum = 0, cnt = 0;
        for(int i=0; i<nums.size(); i++){
            sum += nums[i];
            if (prefix.count(sum - k)){
                cnt += prefix[sum-k];
            }
            prefix[sum]++;
        }
        return cnt;
    }
};